lib/spack/spack/build_systems/python.py

```diff
@@ -128,191 +112,79 @@ def import_modules(self):
-
-    def install_args(self, spec, prefix):
-        """Arguments to pass to install."""
-        args = ['--prefix={0}'.format(prefix)]
-
-        # This option causes python packages (including setuptools) NOT
-        # to create eggs or easy-install.pth files.  Instead, they
-        # install naturally into $prefix/pythonX.Y/site-packages.
-        #
-        # Eggs add an extra level of indirection to sys.path, slowing
-        # down large HPC runs.  They are also deprecated in favor of
-        # wheels, which use a normal layout when installed.
-        #
-        # Spack manages the package directory on its own by symlinking
-        # extensions into the site-packages directory, so we don't really
-        # need the .pth files or egg directories, anyway.
-        #
-        # We need to make sure this is only for build dependencies. A package
-        # such as py-basemap will not build properly with this flag since
-        # it does not use setuptools to build and those does not recognize
-        # the --single-version-externally-managed flag
-        if ('py-setuptools' == spec.name or          # this is setuptools, or
-            'py-setuptools' in spec._dependencies and  # it's an immediate dep
-            'build' in spec._dependencies['py-setuptools'].deptypes):
-            args += ['--single-version-externally-managed']
-
-        # Get all relative paths since we set the root to `prefix`
-        # We query the python with which these will be used for the lib and inc
-        # directories. This ensures we use `lib`/`lib64` as expected by python.
-        pure_site_packages_dir = spec['python'].package.config_vars[
-            'python_lib']['false']['false']
-        plat_site_packages_dir = spec['python'].package.config_vars[
-            'python_lib']['true']['false']
-        inc_dir = spec['python'].package.config_vars['python_inc']['true']
-
-        args += ['--root=%s' % prefix,
-                 '--install-purelib=%s' % pure_site_packages_dir,
-                 '--install-platlib=%s' % plat_site_packages_dir,
-                 '--install-scripts=bin',
-                 '--install-data=',
-                 '--install-headers=%s' % inc_dir
-                 ]
-
-        return args
-
-    def install_lib(self, spec, prefix):
-        """Install all Python modules (extensions and pure Python)."""
-        args = self.install_lib_args(spec, prefix)
-
-        self.setup_py('install_lib', *args)
-
-    def install_lib_args(self, spec, prefix):
-        """Arguments to pass to install_lib."""
-        return []
 
-    def install_headers(self, spec, prefix):
-        """Install C/C++ header files."""
-        args = self.install_headers_args(spec, prefix)
-
-        self.setup_py('install_headers', *args)
-
-    def install_headers_args(self, spec, prefix):
-        """Arguments to pass to install_headers."""
-        return []
+        args = PythonPackage._std_args(self) + ['--prefix=' + prefix]
```

Comment:
I notice the _std_args getting pulled out in a lot of places like this, might it be worth using `add_default_arg` on the pip executable for all of these?  I could see someone wanting to use a different set, so perhaps not, but the common case seems to be `pip(*std_args,*args)` more or less, so having a less cumbersome way to do that seems potentially beneficial.

Reply:
I like the idea. The only reason I hesitate to do that is because we might want to use `pip` without the `install` arg. For example, someone suggested using `pip check` to ensure that there are no missing deps after an installation (will prob add this check in a follow-up PR).

Reply:
That makes sense.  It actually makes me wish we had a way to have the Executable have default arguments that are ignored when non-keyword arguments are supplied, but used either when called with no arguments or when called with a keyword argument to extend the set with.  Or something like that to make it easy.  Even just an extra copy of the Executable object as say, `pip_install_std()`, that has the default arguments in it so the extras are added might get us there, just something that seems worth considering.

Reply:
We could definitely move the global options to `add_default_arg` and keep the install options as `std_install_args`.

Reply:
Actually, had one further thought on this.  What about either a new method on `PythonPackage` for `extra_pip_arguments` or a keyword argument to `PythonPackage.install()`?  I could throw either up as a follow-on to keep from having to refer to the default set explicitly everywhere.

Reply:
Do you have a use case for this yet? I can definitely imagine needing to add it someday but would like to see a use case before adding more complexity than we need.

Reply:
I suppose no more than we have for the `install_options` method, which given that it's a pip-based class now I would actually expect to do what I'm referring to rather than what it actually does.  It's a relatively common thing to need, and seems unfortunate to need to override `install` and expand a list the packager otherwise doesn't need to know about, and isn't listed in the default create template, to get it.  As with any of these things, it's syntax sugar of a sort. 🤷

Reply:
For `install_options`, we already have examples where this is needed in dozens of packages. For pip flags do you have an example? Since we only use pip to build a single wheel and copy it to an installation prefix, most of the install flags aren't relevant.